File: 2020/D18/d18.py
"""
Advent Of Code 2020
Day 18: Operation Order part 1 & 2

https://adventofcode.com/2020/day/18
"""
from time import perf_counter as pfc
import os

def readInput(filename):
  with open(os.path.join("2020", "D18", filename)) as f:
    lines = f.readlines()
  input = (tuple(d.strip('\n').replace(" ", "")) for d in lines)
  return tuple(input)

def calculate(operands, operators):
  op1 = operands.pop()
  op2 = operands.pop()
  oper = operators.pop()
  return op1 + op2 if oper == '+' else op1 * op2

def part1(input):
  startTime = pfc()
  result = 0
  for exp in input:
    if not exp:
      continue
    operators = []
    operands = []
    for s in exp:
      if s in "0123456789":
        operands.append(int(s))
        if operators and operators[-1] in "+*":
          r = calculate(operands, operators)
          operands.append(r)
      elif s in "+*(":
        operators.append(s)
      elif s == ')':
        while operators and operators[-1] != '(':
          r = calculate(operands, operators)
          operands.append(r)
        if operators:  # pop and discard the left parentesis
          operators.pop()
        while operators and operators[-1] in "+*":
          r = calculate(operands, operators)
          operands.append(r)
    result += operands[0]

  print(f"Part 1 time: {pfc() - startTime:.4}")
  return result

def toPostfix(exp):
  postfix = []
  stack = []
  for s in exp:
    if s in "0123456789":
      postfix.append(int(s))
    elif s == "(":
      stack.append(s)
    elif s == ")":
      while stack and stack[-1] != "(":
        postfix.append(stack.pop())
      if stack:
        stack.pop()
    elif s in "+*":
      if not stack or stack[-1] == "(":
        stack.append(s)
      else:
        while stack and stack[-1] not in "(*":
          postfix.append(stack.pop())
        stack.append(s)
  while stack:
    postfix.append(stack.pop())
  return postfix

def part2(input):
  startTime = pfc()
  result = 0
  for exp in input:
    if not exp:
      continue
    postfix = toPostfix(exp)
    # evaluate postfix exp
    stack = []
    for s in postfix:
      if isinstance(s, int):
        stack.append(s)
      else:
        r = calculate(stack, [s])
        stack.append(r)
    result += stack[0]

  print(f"Part 2 time: {pfc() - startTime:.4}")
  return result

#input = readInput(f"d18-sample.txt")
input = readInput(f"d18-input.txt")
print(f"Part 1 for puzzle input: {part1(input)}")
print(f"Part 2 for puzzle input: {part2(input)}")

File: 2024/D18/d18.js
/*
Advent Of Code 2024
Day 18: RAM Run part 1 & 2

https://adventofcode.com/2024/day/18
*/

const fs = require('fs');
const path = require('path');

const YEAR = '2024';
const DAY = '18';

function readInput(filename) {
  const readRawInput = () => {
    const data = fs.readFileSync(path.join(`${YEAR}`, `D${DAY}`, filename));
    const input = data
      .toString()
      .split('\n')
      .filter((l) => l.length > 0)
      .map((l) => l.split(',').map((v) => parseInt(v)));
    return input;
  };
  try {
    return readRawInput();
  } catch (err) {
    console.error(err);
  }
}

const DIR4 = [
  [0, -1],
  [1, 0],
  [0, 1],
  [-1, 0],
];

function getNeighborPositions(input, x, y, dir) {
  const height = input.length;
  const width = input[0].length;
  const pos = dir.map(([accX, accY]) => ({ x: x + accX, y: y + accY }));
  return pos.filter(({ x, y }) => !(x < 0 || x >= width || y < 0 || y >= height));
}

const parseMapXY = (input, actionFct) => {
  const height = input.length;
  const width = input[0].length;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const r = actionFct(x, y, input[y][x], width, height);
      if (!r) continue;
      if (r.break) break;
      if (r.result) return result;
    }
  }
};

function solution1(input, width, height, fallByteMaxCount) {
  const map = new Array(height).fill('.').map((r) => new Array(width).fill('.'));

  for (let i = 0; i < fallByteMaxCount; i++) {
    const [x, y] = input[i];
    map[y][x] = '#';
  }

  const maze = new Map();
  const updateMaze = (x, y, v) => {
    const k = `${x}:${y}`;
    if (!maze.has(k)) maze.set(k, { x, y });
    maze.set(k, { ...maze.get(k), ...v });
  };
  const getMazePos = (x, y) => maze.get(`${x}:${y}`);

  parseMapXY(map, (x, y, elt) => {
    if (elt === '.') updateMaze(x, y, { step: Infinity, visited: false, prev: { x: -1, y: -1 } });
  });

  const startPos = { x: 0, y: 0 };
  const endPos = { x: width - 1, y: height - 1 };
  updateMaze(startPos.x, startPos.y, { visited: true, step: 0 });

  // shortest path
  const candidates = [getMazePos(startPos.x, startPos.y)];
  while (candidates.length > 0) {
    candidates.sort((a, b) => b.step - a.step);
    const { x, y, step } = candidates.pop();

    updateMaze(x, y, { visited: true });
    const neighbors = getNeighborPositions(map, x, y, DIR4);
    const validNeighbors = neighbors.filter(({ x, y }) => map[y][x] !== '#' && !getMazePos(x, y).visited);
    for (const n of validNeighbors) {
      const nPos = getMazePos(n.x, n.y);
      if (step + 1 < nPos.step) {
        nPos.step = step + 1;
        nPos.prev = { x: x, y: y };
        candidates.push(nPos);
      }
    }
  }

  let result = getMazePos(endPos.x, endPos.y).step;
  return result;
}

function solution2(input, width, height, fallByteMaxCount) {
  const map = new Array(height).fill('.').map((r) => new Array(width).fill('.'));

  for (let i = 0; i < fallByteMaxCount; i++) {
    const [x, y] = input[i];
    map[y][x] = '#';
  }

  const maze = new Map();
  const updateMaze = (x, y, v) => {
    const k = `${x}:${y}`;
    if (!maze.has(k)) maze.set(k, { x, y });
    maze.set(k, { ...maze.get(k), ...v });
  };
  const getMazePos = (x, y) => maze.get(`${x}:${y}`);
  const resetMaze = () => {
    for (k of maze.keys()) maze.set(k, { ...maze.get(k), ...{ step: Infinity, visited: false, prev: { x: -1, y: -1 } } });
  };

  parseMapXY(map, (x, y, elt) => {
    if (elt === '.') updateMaze(x, y, { step: Infinity, visited: false, prev: { x: -1, y: -1 } });
  });

  let byteIndex = fallByteMaxCount;
  while (true) {
    // new block
    const [bx, by] = input[byteIndex];
    map[by][bx] = '#';
    maze.delete(`${bx}:${by}`);
    resetMaze();

    const startPos = { x: 0, y: 0 };
    const endPos = { x: width - 1, y: height - 1 };
    updateMaze(startPos.x, startPos.y, { visited: true, step: 0 });

    // shortest path
    const candidates = [getMazePos(startPos.x, startPos.y)];
    while (candidates.length > 0) {
      candidates.sort((a, b) => b.step - a.step);
      const { x, y, step } = candidates.pop();
      updateMaze(x, y, { visited: true });
      const neighbors = getNeighborPositions(map, x, y, DIR4);
      const validNeighbors = neighbors.filter(({ x, y }) => map[y][x] !== '#' && !getMazePos(x, y).visited);
      for (const n of validNeighbors) {
        const nPos = getMazePos(n.x, n.y);
        if (step + 1 < nPos.step) {
          nPos.step = step + 1;
          nPos.prev = { x: x, y: y };
          candidates.push(nPos);
        }
      }
    }

    if (getMazePos(endPos.x, endPos.y).step === Infinity) break;
    byteIndex++;
    if (byteIndex >= input.length) break;
  }

  if (byteIndex >= input.length) {
    console.log("cloud't find a blocking path");
    return 0;
  }

  let result = `${input[byteIndex][0]},${input[byteIndex][1]}`;
  return result;
}

function part1(mode) {
  const input = readInput(`d${DAY}-${mode}.txt`);
  const width = mode === 'sample' ? 7 : 71;
  const height = mode === 'sample' ? 7 : 71;
  const fallByteMaxCount = mode === 'sample' ? 12 : 1024;
  const startTime = performance.now();
  const result = solution1(input, width, height, fallByteMaxCount);
  const elapsed = performance.now() - startTime;
  console.log(`Part 1 ${mode} time: ${Math.floor(elapsed / 60000)}:${Math.floor((elapsed % 60000) / 1000)}:${Math.floor(elapsed % 1000)}`);
  console.log(`Part 1 ${mode} result: ${result}`);
}

function part2(mode) {
  const input = readInput(`d${DAY}-${mode}.txt`);
  const width = mode === 'sample' ? 7 : 71;
  const height = mode === 'sample' ? 7 : 71;
  const fallByteMaxCount = mode === 'sample' ? 12 : 1024;
  const startTime = performance.now();
  const result = solution2(input, width, height, fallByteMaxCount);
  const elapsed = performance.now() - startTime;
  console.log(`Part 2 ${mode} time: ${Math.floor(elapsed / 60000)}:${Math.floor((elapsed % 60000) / 1000)}:${Math.floor(elapsed % 1000)}`);
  console.log(`Part 2 ${mode} result: ${result}`);
}

part1('sample');
part1('input');
part2('sample');
part2('input');
